File: 1707-maximum-xor-with-an-element-from-array/1707-maximum-xor-with-an-element-from-array.js
/**
 * @param {number[]} nums
 * @param {number[][]} queries
 * @return {number[]}
 */
var maximizeXor = function(nums, queries) {

    // 字典树
    const trie = [null, null];
    const HIGH_BIT = 30;

    function insert(val) {
        // use node to keep track of the trie
        let node = trie;
        for (let i = HIGH_BIT - 1; i >= 0; --i) {
            const bit = (val >> i) & 1;
            if (node[bit] === null) {
                node[bit] = [null, null];
            }
            node = node[bit];
        }
    }

    function getMaxXor(val) {
        let ans = 0;
        let node = trie;
        for (let i = HIGH_BIT - 1; i >= 0; --i) {
            let bit = (val >> i) & 1;
            if (node[bit ^ 1] !== null) {
                ans |= 1 << i;
                bit ^= 1;
            }
            node = node[bit];
        }
        return ans;
    }

    queries.forEach((v, i) => v.push(i));
    queries.sort((a, b) => a[1] - b[1]);
    const qlen = queries.length;
    const ans = new Array(qlen);
    let idx = 0;

    nums.sort((a, b) => a - b);
    const n = nums.length;
    for (const q of queries) {
        const x = q[0];
        const m = q[1];
        const qid = q[2];

        while (idx < n && nums[idx] <= m) {
            insert(nums[idx]);
            ++idx;
        }
        if (idx === 0) {
            ans[qid] = -1;
        } else {
            ans[qid] = getMaxXor(x);
        }
    }

    return ans;
};

File: 1707-maximum-xor-with-an-element-from-array/1707-maximum-xor-with-an-element-from-array.py
class Trie:
    def __init__(self):
        self.root = {}
        
    def insert(self, num):
        node = self.root
        for i in range(31, -1, -1):
            bit = (num >> i) & 1
            if bit not in node:
                node[bit] = {}
            node = node[bit]
            
    def query(self, num):        
        node, ans = self.root, 0
        for i in range(31, -1, -1):
            bit = (num >> i) & 1
            if not node:
                return -1
            if 1-bit in node:
                ans |= (1 << i)
                node = node[1-bit]
            else:
                node = node[bit]            
        return ans
              

class Solution:
    def maximizeXor(self, nums: List[int], queries: List[List[int]]) -> List[int]:
        nums.sort()
        queries = sorted(enumerate(queries), key=lambda x:x[1][1])
        trie = Trie()
        ans = [-1] * len(queries)
        j = 0
        for i, (x, m) in queries:
            while j < len(nums) and nums[j] <= m:
                trie.insert(nums[j])
                j += 1
            ans[i] = trie.query(x)
        return ans